File: fcr-calculator/calculator.py
import pandas as pd
import numpy as np
from dataclasses import dataclass


@dataclass
class BatteryConfig:
    power_mw: float
    capacity_mwh: float
    efficiency: float = 0.90
    soc_min: float = 0.20
    soc_max: float = 0.80


@dataclass
class SimulationResult:
    hourly_df: pd.DataFrame
    total_revenue_eur: float
    available_hours: int
    total_hours: int
    availability_pct: float
    avg_price_eur: float


def calculate_fcr_n_activation(frequency: float, power_mw: float) -> float:
    """
    Calculate FCR-N power activation based on frequency.

    FCR-N activates linearly in 49.9-50.1 Hz band:
    - At 49.9 Hz: full discharge (+power_mw)
    - At 50.0 Hz: zero activation
    - At 50.1 Hz: full charge (-power_mw)

    Returns power in MW (positive = discharge, negative = charge).
    """
    if frequency <= 49.9:
        return power_mw
    elif frequency >= 50.1:
        return -power_mw
    else:
        return (50.0 - frequency) / 0.1 * power_mw


def simulate_soc_hourly(
    freq_df: pd.DataFrame,
    config: BatteryConfig,
    start_soc: float = 0.5
) -> pd.DataFrame:
    """
    Simulate battery SOC evolution using 1-second frequency data.

    Returns DataFrame with hourly SOC statistics and availability.
    """
    freq_df = freq_df.copy()
    freq_df["hour"] = freq_df["timestamp"].dt.floor("h")

    usable_capacity = config.capacity_mwh * (config.soc_max - config.soc_min)
    min_energy = config.capacity_mwh * config.soc_min
    max_energy = config.capacity_mwh * config.soc_max

    results = []
    current_energy = config.capacity_mwh * start_soc

    for hour, group in freq_df.groupby("hour"):
        hour_start_energy = current_energy
        unavailable_seconds = 0

        for _, row in group.iterrows():
            freq = row["grid_frequency"]
            power = calculate_fcr_n_activation(freq, config.power_mw)

            # Energy change per second (power in MW, time in hours)
            delta_energy = power / 3600  # MWh per second

            # Apply efficiency loss
            if delta_energy > 0:  # Discharging
                delta_energy /= np.sqrt(config.efficiency)
            else:  # Charging
                delta_energy *= np.sqrt(config.efficiency)

            new_energy = current_energy - delta_energy

            # Check SOC limits
            if new_energy < min_energy or new_energy > max_energy:
                unavailable_seconds += 1
                # Clamp to limits
                new_energy = np.clip(new_energy, min_energy, max_energy)

            current_energy = new_energy

        hour_end_energy = current_energy
        available = unavailable_seconds < 60  # Available if <1 minute at limits

        results.append({
            "hour": hour,
            "soc_start": hour_start_energy / config.capacity_mwh,
            "soc_end": hour_end_energy / config.capacity_mwh,
            "soc_change": (hour_end_energy - hour_start_energy) / config.capacity_mwh,
            "unavailable_seconds": unavailable_seconds,
            "available": available
        })

    return pd.DataFrame(results)


def calculate_revenue(
    price_df: pd.DataFrame,
    soc_df: pd.DataFrame,
    config: BatteryConfig
) -> SimulationResult:
    """
    Calculate revenue by combining price data with availability from SOC simulation.
    """
    # Merge price and SOC data
    merged = price_df.merge(
        soc_df,
        left_on=price_df["timestamp"].dt.floor("h"),
        right_on="hour",
        how="left"
    )

    # If no frequency data for an hour, assume available
    merged["available"] = merged["available"].fillna(True)

    # Calculate hourly revenue
    merged["revenue_eur"] = merged.apply(
        lambda row: config.power_mw * row["FCR-N Price EUR/MW"] if row["available"] else 0,
        axis=1
    )

    total_revenue = merged["revenue_eur"].sum()
    available_hours = merged["available"].sum()
    total_hours = len(merged)
    availability_pct = available_hours / total_hours * 100 if total_hours > 0 else 0
    avg_price = merged["FCR-N Price EUR/MW"].mean()

    result_df = merged[[
        "timestamp", "FCR-N Price EUR/MW", "available", "revenue_eur",
        "soc_start", "soc_end"
    ]].copy()

    return SimulationResult(
        hourly_df=result_df,
        total_revenue_eur=total_revenue,
        available_hours=int(available_hours),
        total_hours=total_hours,
        availability_pct=availability_pct,
        avg_price_eur=avg_price
    )


def calculate_simple_revenue(
    price_df: pd.DataFrame,
    power_mw: float,
    availability_pct: float = 100.0
) -> SimulationResult:
    """
    Calculate revenue without SOC simulation (simple availability factor).
    """
    df = price_df.copy()
    factor = availability_pct / 100.0

    df["available"] = True
    df["revenue_eur"] = power_mw * df["FCR-N Price EUR/MW"] * factor

    total_revenue = df["revenue_eur"].sum()
    total_hours = len(df)

    return SimulationResult(
        hourly_df=df[["timestamp", "FCR-N Price EUR/MW", "available", "revenue_eur"]],
        total_revenue_eur=total_revenue,
        available_hours=int(total_hours * factor),
        total_hours=total_hours,
        availability_pct=availability_pct,
        avg_price_eur=df["FCR-N Price EUR/MW"].mean()
    )

File: fcr-calculator-electron/src/renderer/calculator.js
// Battery configuration
class BatteryConfig {
  constructor(powerMw, capacityMwh, efficiency = 0.90, socMin = 0.20, socMax = 0.80) {
    this.powerMw = powerMw;
    this.capacityMwh = capacityMwh;
    this.efficiency = efficiency;
    this.socMin = socMin;
    this.socMax = socMax;
  }
}

// Calculate FCR-N power activation based on frequency
// FCR-N activates linearly in 49.9-50.1 Hz band:
// - At 49.9 Hz: full discharge (+powerMw)
// - At 50.0 Hz: zero activation
// - At 50.1 Hz: full charge (-powerMw)
function calculateFcrActivation(frequency, powerMw) {
  if (frequency <= 49.9) return powerMw;
  if (frequency >= 50.1) return -powerMw;
  return (50.0 - frequency) / 0.1 * powerMw;
}

// Simulate battery SOC evolution using 1-second frequency data with NEM
// NEM (Normal State Energy Management) adds power offset to restore SOC
// while still participating in FCR when frequency is in normal band
function simulateSocHourly(freqData, config, startSoc = 0.5) {
  console.log(`[SOC] Starting simulation with ${freqData.length} samples`);

  const minEnergy = config.capacityMwh * config.socMin;
  const maxEnergy = config.capacityMwh * config.socMax;

  // NEM parameters
  const NEM_POWER_RATIO = 0.34; // 34% of capacity for NEM
  const NEM_WINDOW_SECONDS = 120; // 2-minute rolling average (faster response)

  // NEM SOC thresholds - relative to user's SOC limits with 15% buffer
  const socRange = config.socMax - config.socMin;
  const nemBuffer = socRange * 0.25; // Start NEM when 25% into the usable range
  const nemEnableLower = config.socMin + nemBuffer;
  const nemDisableLower = config.socMin + socRange * 0.5; // Disable at midpoint
  const nemEnableUpper = config.socMax - nemBuffer;
  const nemDisableUpper = config.socMax - socRange * 0.5;

  console.log(`[SOC] NEM thresholds: enable=${(nemEnableLower * 100).toFixed(1)}%/${(nemEnableUpper * 100).toFixed(1)}%, disable=${(nemDisableLower * 100).toFixed(1)}%/${(nemDisableUpper * 100).toFixed(1)}%`);

  // freqData is now { frequencies: Float64Array, summary, startTime, hours }
  const frequencies = freqData.frequencies;
  const nSamples = frequencies.length;
  const nHours = Math.floor(nSamples / 3600);
  const startTimeMs = freqData.startTime.getTime();
  console.log(`[SOC] Processing ${nHours} hours (${nSamples} samples)`);

  const results = [];
  let currentEnergy = config.capacityMwh * startSoc;
  const sqrtEfficiency = Math.sqrt(config.efficiency);

  // NEM state tracking
  const nemAllowedHistory = [];
  let nemActiveLow = false;
  let nemActiveHigh = false;

  // Stats tracking
  let totalNemActiveSeconds = 0;
  let totalNemChargingSeconds = 0;
  let totalNemDischargingSeconds = 0;
  let minSoc = 1;
  let maxSoc = 0;

  console.log('[SOC] Processing hours with NEM...');
  for (let h = 0; h < nHours; h++) {
    const hourStartEnergy = currentEnergy;
    const hourTimestamp = new Date(startTimeMs + h * 3600000);
    let unavailableSeconds = 0;

    // Process 3600 samples for this hour
    const hourStart = h * 3600;
    const hourEnd = hourStart + 3600;
    for (let i = hourStart; i < hourEnd; i++) {
      const freq = frequencies[i];
      const soc = currentEnergy / config.capacityMwh;
      const inNormalBand = freq > 49.9 && freq < 50.1;

      // Calculate NEM_Allowed based on SOC and frequency band
      let nemAllowed = 0;
      if (inNormalBand) {
        // Low SOC - need to charge (nemAllowed = -1)
        if (soc < nemEnableLower) {
          nemActiveLow = true;
        } else if (soc >= nemDisableLower) {
          nemActiveLow = false;
        }

        // High SOC - need to discharge (nemAllowed = +1)
        if (soc > nemEnableUpper) {
          nemActiveHigh = true;
        } else if (soc <= nemDisableUpper) {
          nemActiveHigh = false;
        }

        if (nemActiveLow) {
          nemAllowed = -1; // Charge to restore SOC
          totalNemChargingSeconds++;
          totalNemActiveSeconds++;
        } else if (nemActiveHigh) {
          nemAllowed = 1; // Discharge to restore SOC
          totalNemDischargingSeconds++;
          totalNemActiveSeconds++;
        }
      }

      // Track SOC range
      if (soc < minSoc) minSoc = soc;
      if (soc > maxSoc) maxSoc = soc;

      // Rolling 2-min average for NEM_current
      nemAllowedHistory.push(nemAllowed);
      if (nemAllowedHistory.length > NEM_WINDOW_SECONDS) {
        nemAllowedHistory.shift();
      }
      let nemSum = 0;
      for (let i = 0; i < nemAllowedHistory.length; i++) {
        nemSum += nemAllowedHistory[i];
      }
      const nemCurrent = nemSum / nemAllowedHistory.length;

      // FCR power response
      const fcrPower = calculateFcrActivation(freq, config.powerMw);

      // NEM power offset (adds to FCR)
      const nemPower = NEM_POWER_RATIO * config.powerMw * nemCurrent;

      // Total power = FCR + NEM
      const totalPower = fcrPower + nemPower;

      // Energy change per second
      let deltaEnergy = totalPower / 3600;

      // Apply efficiency
      if (deltaEnergy > 0) {
        deltaEnergy /= sqrtEfficiency;
      } else {
        deltaEnergy *= sqrtEfficiency;
      }

      let newEnergy = currentEnergy - deltaEnergy;

      // Check hard SOC limits
      if (newEnergy < minEnergy || newEnergy > maxEnergy) {
        unavailableSeconds++;
        newEnergy = Math.max(minEnergy, Math.min(maxEnergy, newEnergy));
      }

      currentEnergy = newEnergy;
    }

    const hourEndEnergy = currentEnergy;
    const available = unavailableSeconds < 60;

    results.push({
      hour: hourTimestamp,
      socStart: hourStartEnergy / config.capacityMwh,
      socEnd: hourEndEnergy / config.capacityMwh,
      socChange: (hourEndEnergy - hourStartEnergy) / config.capacityMwh,
      unavailableSeconds,
      available
    });
  }

  // Log summary stats
  const totalSeconds = nHours * 3600;
  const unavailableHours = results.filter(r => !r.available).length;
  console.log('[SOC] ═══════════════════════════════════════');
  console.log(`[SOC] SOC Range: ${(minSoc * 100).toFixed(1)}% - ${(maxSoc * 100).toFixed(1)}%`);
  console.log(`[SOC] NEM Active: ${(totalNemActiveSeconds / totalSeconds * 100).toFixed(1)}% of time`);
  console.log(`[SOC]   - Charging (low SOC): ${(totalNemChargingSeconds / totalSeconds * 100).toFixed(1)}%`);
  console.log(`[SOC]   - Discharging (high SOC): ${(totalNemDischargingSeconds / totalSeconds * 100).toFixed(1)}%`);
  console.log(`[SOC] Unavailable hours: ${unavailableHours} / ${results.length} (${(unavailableHours / results.length * 100).toFixed(1)}%)`);
  console.log('[SOC] ═══════════════════════════════════════');

  return results;
}

// Calculate revenue by combining price data with availability from SOC simulation
function calculateRevenue(priceData, socData, config) {
  // Create lookup map for SOC data by hour
  const socByHour = new Map();
  for (const row of socData) {
    socByHour.set(row.hour.getTime(), row);
  }

  const hourlyResults = [];
  let totalRevenue = 0;
  let availableHours = 0;

  for (const priceRow of priceData) {
    const hourKey = new Date(priceRow.timestamp).setMinutes(0, 0, 0);
    const socRow = socByHour.get(hourKey);

    const available = socRow ? socRow.available : true;
    const revenue = available ? config.powerMw * priceRow.price : 0;

    totalRevenue += revenue;
    if (available) availableHours++;

    hourlyResults.push({
      timestamp: priceRow.timestamp,
      price: priceRow.price,
      available,
      revenue,
      socStart: socRow?.socStart ?? null,
      socEnd: socRow?.socEnd ?? null
    });
  }

  const totalHours = priceData.length;
  const avgPrice = priceData.reduce((sum, r) => sum + r.price, 0) / totalHours;

  return {
    hourlyData: hourlyResults,
    totalRevenue,
    availableHours,
    totalHours,
    availabilityPct: totalHours > 0 ? (availableHours / totalHours) * 100 : 0,
    avgPrice
  };
}

// Calculate revenue without SOC simulation (simple availability factor)
function calculateSimpleRevenue(priceData, powerMw, availabilityPct = 100) {
  const factor = availabilityPct / 100;

  const hourlyResults = [];
  let totalRevenue = 0;

  for (const row of priceData) {
    const revenue = powerMw * row.price * factor;
    totalRevenue += revenue;

    hourlyResults.push({
      timestamp: row.timestamp,
      price: row.price,
      available: true,
      revenue
    });
  }

  const totalHours = priceData.length;
  const avgPrice = priceData.reduce((sum, r) => sum + r.price, 0) / totalHours;

  return {
    hourlyData: hourlyResults,
    totalRevenue,
    availableHours: Math.round(totalHours * factor),
    totalHours,
    availabilityPct,
    avgPrice
  };
}

// Export for use in app.js
window.Calculator = {
  BatteryConfig,
  calculateFcrActivation,
  simulateSocHourly,
  calculateRevenue,
  calculateSimpleRevenue
};
